docs/busy-throttling.md

```diff
@@ -0,0 +1,185 @@
+# Throttling on Busy Responses
+
+In TChannel, a busy response is sent to indicate that a service is overloaded. Upon
+receiving busy responses, the client should reduce its requests to the service so
+that the service can recover to a healthy state.
+
+## Goals
+There are three goals we want to achieve:
+
+* When a busy response is received, we want to exponentially reduce the outgoing requests
+  to the service.
+* The longer the busy responses are received, the longer it should take to recover the outgoing
+  requests. This is to reduce the flappy scenarios where the service is back and forth in good
+  and bad states.
+* When there is no busy reponses and the service is back to healthy. Traffic should not take too
+  long to recover.
+
+
+## Design
+
+In order to do throttling, it is necessary to measure how fast the client is sending
+the requests. We use RPS (Request Per Second), i.e., the number of request sent every
+second. Given the client `clientA` that sends requests to the sevice `serviceB` with a
+RPS of `RpsToB` and a RPS limit of `RpsLimitToB`, there are three scenairos that we
+care about:
+
+1. All requests responeded without any busy responses. In this case, no request
+   should be throttled. `clientA` should continue to send all requests to `serviceB`. 
+2. If a busy response is received, `clientA` should throttle its sent requests to `serviceB`.
+   We use an exponential throttling rate, i.e., for every busy response received,
+   we set the RPS limit `RpsLimitToB` to `RpsToB / 2`. This will
+   actively cut down half of the request volume from `clientA` to `serviceB`.
+3. When requests to `serviceB` is under throttling, there should be a gradual increase
+   of the RPS limit to allow some requests coming through. Based on the simulation, we
+   set the recover rate as 3, i.e., for every second, we set `RpsLimitToB` = `RpsLimitToB * 3` .
+
+For throttling on busy responses, it is important that normal traffic is not affected. When
+there is a busy response, requests will get throttled in an exponential rate. On the other 
+hand, the recover rate makes sure that the throttling condition can be gradually lifted when
+the service's throughput is improved.  
+
+## Simulation
+We use a simulation program to validate the design. The program simulates time as a incremental counter of
+seconds. For each second, there will be a customizable amout of out requests and busy responses. It is 
+important to make sure that the number of busy responses never exceeds the number of actual out requests.
+
+First, We let the simulation run for 10 seconds (i.e., 10 increment of time counter) with various traffic
```

Comment:
That seems an awfully limited test scenario, I'd like to at least see adding a "normal" starting phase before the troubled time of busies.

Even better would be to simulate with a slowly varying (e.g. over the course of an hour(s) or even up to a day) and rate-of-request triggered busies in the endpoint.